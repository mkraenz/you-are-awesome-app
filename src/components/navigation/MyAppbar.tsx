import React, { FC } from "react";
import { Appbar } from "react-native-paper";

interface Props {
    title: string;
    icon?: string;
}

const MyAppbar: FC<Props> = ({ title, icon = "menu" }) => {
    return (
        <Appbar.Header>
            <Appbar.Action icon={icon} onPress={() => undefined} />
            <Appbar.Content title={title} />
            <Appbar.Action icon="magnify" onPress={() => undefined} />
            <Appbar.Action icon="dots-vertical" onPress={() => undefined} />
        </Appbar.Header>
    );
};

export default MyAppbar;
