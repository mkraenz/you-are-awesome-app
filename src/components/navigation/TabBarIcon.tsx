import { FontAwesome } from "@expo/vector-icons";
import React, { FC } from "react";
import { StyleSheet } from "react-native";
import { useTheme } from "react-native-paper";

const styles = StyleSheet.create({
    icon: {
        marginBottom: -3,
    },
});

interface Props {
    name: string;
    focused: boolean;
}

const TabBarIcon: FC<Props> = ({ name, focused }) => {
    const { colors } = useTheme();

    return (
        <FontAwesome
            name={name}
            size={26}
            style={styles.icon}
            color={focused ? colors.accent : colors.disabled}
        />
    );
};

export default TabBarIcon;
