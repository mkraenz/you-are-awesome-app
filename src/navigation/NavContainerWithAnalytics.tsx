import {
    NavigationContainer,
    NavigationContainerRef,
} from "@react-navigation/native";
import * as Analytics from "expo-firebase-analytics";
import React, { FC } from "react";
import { useTheme } from "react-native-paper";
import { CONFIG } from "../config";
import { FullTheme } from "../themes/theme";

/** following https://reactnavigation.org/docs/screen-tracking/ */
const NavContainerWithAnalytics: FC = ({ children }) => {
    const routeNameRef = React.useRef<string>(null);
    const navigationRef = React.useRef<NavigationContainerRef>(null);
    const theme = useTheme() as FullTheme;

    Analytics.setDebugModeEnabled(CONFIG.debugAnalytics);
    return (
        <NavigationContainer
            theme={theme}
            ref={navigationRef}
            onReady={() => {
                const currentRoute = navigationRef.current!.getCurrentRoute();
                (routeNameRef as any).current = currentRoute!.name;
            }}
            onStateChange={() => {
                const previousRouteName = routeNameRef.current;
                const currentRouteName = navigationRef.current!.getCurrentRoute()!
                    .name;

                if (previousRouteName !== currentRouteName) {
                    Analytics.setCurrentScreen(currentRouteName);
                }

                (routeNameRef as any).current = currentRouteName;
            }}
        >
            {children}
        </NavigationContainer>
    );
};

export default NavContainerWithAnalytics;
