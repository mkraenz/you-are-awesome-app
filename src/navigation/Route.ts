export enum Route {
    Home = "Home",
    Settings = "Settings",
    Contribute = "Contribute",
}
