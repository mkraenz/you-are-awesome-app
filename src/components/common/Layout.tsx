import React, { FC } from "react";
import { useTranslation } from "react-i18next";
import { StyleSheet, View } from "react-native";
import { Surface, useTheme } from "react-native-paper";
import { Route } from "../../navigation/Route";
import MyAppbar from "../navigation/MyAppbar";

const styles = StyleSheet.create({
    container: {
        width: "100%",
        height: "100%",
    },
    contentContainer: {
        flex: 1,
        paddingHorizontal: 16,
    },
});

interface Props {
    route: Route;
}

const Layout: FC<Props> = (props) => {
    const { t } = useTranslation();
    const theme = useTheme();
    const containerStyles = {
        ...styles.container,
        backgroundColor: theme.dark
            ? theme.colors.background
            : theme.colors.surface,
    };

    return (
        <Surface style={containerStyles}>
            <MyAppbar title={t(props.route)} />
            <View style={styles.contentContainer}>{props.children}</View>
        </Surface>
    );
};
export default Layout;
