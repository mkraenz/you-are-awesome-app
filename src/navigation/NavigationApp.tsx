import { createMaterialBottomTabNavigator } from "@react-navigation/material-bottom-tabs";
import { NavigationContainer } from "@react-navigation/native";
import React from "react";
import { useTranslation } from "react-i18next";
import { useTheme } from "react-native-paper";
import TabBarIcon from "../components/navigation/TabBarIcon";
import { START_SCREEN } from "../config";
import ContributionScreen from "../screens/ContributionScreen";
import HomeScreen from "../screens/HomeScreen";
import SettingsScreen from "../screens/SettingsScreen";
import { Route } from "./Route";

const Tab = createMaterialBottomTabNavigator();
const NavigationApp = () => {
    const theme = useTheme();
    const { t } = useTranslation();
    return (
        <NavigationContainer theme={theme as any}>
            <Tab.Navigator initialRouteName={t(START_SCREEN)}>
                <Tab.Screen
                    name={t(Route.Contribute)}
                    component={ContributionScreen}
                    options={{
                        tabBarIcon: ({ focused }) => (
                            <TabBarIcon focused={focused} name="share-alt" />
                        ),
                    }}
                />
                <Tab.Screen
                    name={t(Route.Home)}
                    component={HomeScreen}
                    options={{
                        tabBarIcon: ({ focused }) => (
                            <TabBarIcon focused={focused} name="home" />
                        ),
                    }}
                />
                <Tab.Screen
                    name={t(Route.Settings)}
                    component={SettingsScreen}
                    options={{
                        tabBarIcon: ({ focused }) => (
                            <TabBarIcon focused={focused} name="cog" />
                        ),
                    }}
                />
            </Tab.Navigator>
        </NavigationContainer>
    );
};

export default NavigationApp;
