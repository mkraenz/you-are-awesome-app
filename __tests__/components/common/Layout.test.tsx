import React from "react";
import "react-native";
import { DefaultTheme, Provider as PaperProvider } from "react-native-paper";
// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
import Layout from "../../../src/components/common/Layout";
import { Route } from "../../../src/navigation/Route";
import TestLocalizationProvider from "../../helpers/TestLocalizationProvider";

it("renders correctly", () => {
    const tree = renderer
        .create(
            <PaperProvider theme={DefaultTheme}>
                <TestLocalizationProvider>
                    <Layout route={Route.Home} />
                </TestLocalizationProvider>
            </PaperProvider>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
