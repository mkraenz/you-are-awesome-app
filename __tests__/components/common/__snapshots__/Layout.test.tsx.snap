// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "height": "100%",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 4,
        "width": "100%",
      }
    }
  >
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Array [
          Object {
            "backgroundColor": "#6200ee",
            "elevation": 4,
            "zIndex": 0,
          },
          Object {
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#6200ee",
            "elevation": 0,
            "flexDirection": "row",
            "height": 56,
            "marginTop": 0,
            "paddingHorizontal": 4,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityStates={Array []}
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "margin": 6,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 18,
                  "height": 36,
                  "width": 36,
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "transform": Array [
                      Object {
                        "rotate": "0deg",
                      },
                    ],
                  }
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
                "paddingHorizontal": 12,
              },
              Array [
                Object {
                  "marginLeft": 8,
                },
                false,
                undefined,
              ],
            ]
          }
        >
          <Text
            accessibilityRole="header"
            accessibilityTraits="header"
            accessible={true}
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "fontSize": 17,
                  },
                  undefined,
                ],
              ]
            }
          >
            Home
          </Text>
        </View>
        <View
          accessibilityRole="button"
          accessibilityStates={Array []}
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "margin": 6,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 18,
                  "height": 36,
                  "width": 36,
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "transform": Array [
                      Object {
                        "rotate": "0deg",
                      },
                    ],
                  }
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityStates={Array []}
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "margin": 6,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 18,
                  "height": 36,
                  "width": 36,
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "transform": Array [
                      Object {
                        "rotate": "0deg",
                      },
                    ],
                  }
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTSafeAreaView>
    <View
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 16,
        }
      }
    />
  </View>
</View>
`;
